File: samples/akka-sample-cluster-java/src/main/java/sample/cluster/stats/StatsService.java
package sample.cluster.stats;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import com.fasterxml.jackson.annotation.JsonCreator;
import sample.cluster.CborSerializable;

import java.util.Arrays;
import java.util.List;

public final class StatsService extends AbstractBehavior<StatsService.Command> {

  public interface Command extends CborSerializable {}
  public final static class ProcessText implements Command {
    public final String text;
    public final ActorRef<Response> replyTo;
    public ProcessText(String text, ActorRef<Response> replyTo) {
      this.text = text;
      this.replyTo = replyTo;
    }
  }
  public enum Stop implements Command {
    INSTANCE
  }

  interface Response extends CborSerializable { }
  public static final class JobResult implements Response {
    public final double meanWordLength;
    @JsonCreator
    public JobResult(double meanWordLength) {
      this.meanWordLength = meanWordLength;
    }
    @Override
    public String toString() {
      return "JobResult{" +
          "meanWordLength=" + meanWordLength +
          '}';
    }
  }
  public static final class JobFailed implements Response {
    public final String reason;
    @JsonCreator
    public JobFailed(String reason) {
      this.reason = reason;
    }
    @Override
    public String toString() {
      return "JobFailed{" +
          "reason='" + reason + '\'' +
          '}';
    }
  }

  public static Behavior<Command> create(ActorRef<StatsWorker.Process> workers) {
    return Behaviors.setup(context ->
        new StatsService(context, workers)
    );
  }

  private final ActorRef<StatsWorker.Process> workers;

  private StatsService(ActorContext<Command> context, ActorRef<StatsWorker.Process> workers) {
    super(context);
    this.workers = workers;
  }

  @Override
  public Receive<Command> createReceive() {
    return newReceiveBuilder()
        .onMessage(ProcessText.class, this::process)
        .onMessageEquals(Stop.INSTANCE, () -> Behaviors.stopped())
        .build();
  }

  private Behavior<Command> process(ProcessText command) {
    getContext().getLog().info("Delegating request");
    List<String> words = Arrays.asList(command.text.split(" "));
    getContext().spawnAnonymous(StatsAggregator.create(words, workers, command.replyTo));
    return this;
  }

}

File: samples/akka-sample-cluster-scala/src/main/scala/sample/cluster/stats/StatsService.scala
package sample.cluster.stats

import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Behaviors
import sample.cluster.CborSerializable

import scala.concurrent.duration._

//#service
object StatsService {

  sealed trait Command extends CborSerializable
  final case class ProcessText(text: String, replyTo: ActorRef[Response]) extends Command {
    require(text.nonEmpty)
  }
  case object Stop extends Command

  sealed trait Response extends CborSerializable
  final case class JobResult(meanWordLength: Double) extends Response
  final case class JobFailed(reason: String) extends Response

  def apply(workers: ActorRef[StatsWorker.Process]): Behavior[Command] =
    Behaviors.setup { ctx =>
      // if all workers would crash/stop we want to stop as well
      ctx.watch(workers)

      Behaviors.receiveMessage {
        case ProcessText(text, replyTo) =>
          ctx.log.info("Delegating request")
          val words = text.split(' ').toIndexedSeq
          // create per request actor that collects replies from workers
          ctx.spawnAnonymous(StatsAggregator(words, workers, replyTo))
          Behaviors.same
        case Stop =>
          Behaviors.stopped
      }
    }
}

object StatsAggregator {

  sealed trait Event
  private case object Timeout extends Event
  private case class CalculationComplete(length: Int) extends Event

  def apply(words: Seq[String], workers: ActorRef[StatsWorker.Process], replyTo: ActorRef[StatsService.Response]): Behavior[Event] =
    Behaviors.setup { ctx =>
      ctx.setReceiveTimeout(3.seconds, Timeout)
      val responseAdapter = ctx.messageAdapter[StatsWorker.Processed](processed =>
        CalculationComplete(processed.length)
      )

      words.foreach { word =>
        workers ! StatsWorker.Process(word, responseAdapter)
      }
      waiting(replyTo, words.size, Nil)
    }

  private def waiting(replyTo: ActorRef[StatsService.Response], expectedResponses: Int, results: List[Int]): Behavior[Event] =
    Behaviors.receiveMessage {
      case CalculationComplete(length) =>
        val newResults = results :+ length
        if (newResults.size == expectedResponses) {
          val meanWordLength = newResults.sum.toDouble / newResults.size
          replyTo ! StatsService.JobResult(meanWordLength)
          Behaviors.stopped
        } else {
          waiting(replyTo, expectedResponses, newResults)
        }
      case Timeout =>
        replyTo ! StatsService.JobFailed("Service unavailable, try again later")
        Behaviors.stopped
    }

}
//#service
